fix(ios): Remote issues: #1041 / #1096 onForegroundEvent() and getInitialNotification()

LunatiqueCoder opened the pull request:
### 👋 Hello!

### This PR fixes 
- #912 
- 🤔 Perhaps #999 
- #1041
- #1076
- #1096.
-  #1109 

🙏 Please keep in mind I just learned Objective C and Swift in order to fix those issues.

### What's changed?

- [`onForeground()` is called with `getInitialNotification()`](https://github.com/invertase/notifee/pull/1118/files#diff-ec80c62b93aed0afa4addac8a6cd5fb6534543b3cc39bbd1b706adb9ec82a1ccR97-R98) for `PRESS` and `PRESS_ACTION` events when app is opened from a terminated state.
- [remote payload is picked up when local payload is not available](https://github.com/invertase/notifee/pull/1118/files#diff-e6cee9e1068ee77f7db69b7ee7ccfecefab70ddee393961f93eeeb034d4919afL61-L67). Thanks to this, `getInitialNotification()` will return the correct payload on app start.
- With the above, I think this [code will be unreachable so I removed it](https://github.com/invertase/notifee/pull/1118/files#diff-ec80c62b93aed0afa4addac8a6cd5fb6534543b3cc39bbd1b706adb9ec82a1ccL164-L167)
- `onForeground()`  is called for `DELIVERED` events by [removing this condition](https://github.com/invertase/notifee/pull/1118/files#diff-ec80c62b93aed0afa4addac8a6cd5fb6534543b3cc39bbd1b706adb9ec82a1ccL151-L152), but I have mixed feelings about it. Might need a deeper look.


### How to test
1. Be sure to [configure Notifee to handle remote notifications](https://notifee.app/react-native/docs/ios/remote-notification-support)
2. If you're using [`patch-package`](https://www.npmjs.com/package/patch-package), in your `patches` folder create a file called `@notifee+react-native+9.1.1.patch`
3. Paste the following:


```diff
diff --git a/node_modules/@notifee/react-native/ios/NotifeeCore/NotifeeCore+NSNotificationCenter.m b/node_modules/@notifee/react-native/ios/NotifeeCore/NotifeeCore+NSNotificationCenter.m
index 560a62c..c010367 100644
--- a/node_modules/@notifee/react-native/ios/NotifeeCore/NotifeeCore+NSNotificationCenter.m
+++ b/node_modules/@notifee/react-native/ios/NotifeeCore/NotifeeCore+NSNotificationCenter.m
@@ -58,13 +58,18 @@
 #pragma mark Application Notifications
 
 - (void)application_onDidFinishLaunchingNotification:(nonnull NSNotification *)notification {
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wdeprecated-declarations"
-  UILocalNotification *launchNotification =
-      (UILocalNotification *)notification.userInfo[UIApplicationLaunchOptionsLocalNotificationKey];
-  [[NotifeeCoreUNUserNotificationCenter instance]
-      onDidFinishLaunchingNotification:launchNotification.userInfo];
-  [[NotifeeCoreUNUserNotificationCenter instance] getInitialNotification];
+  NSDictionary *notifUserInfo =
+      notification.userInfo[UIApplicationLaunchOptionsLocalNotificationKey];
+    
+  if (!notifUserInfo) {
+      // Fallback to remote notification key if local notification key is not available
+      notifUserInfo = notification.userInfo[UIApplicationLaunchOptionsRemoteNotificationKey];
+  }
+
+  if (notifUserInfo) {
+      [[NotifeeCoreUNUserNotificationCenter instance] onDidFinishLaunchingNotification:notifUserInfo];
+      [[NotifeeCoreUNUserNotificationCenter instance] getInitialNotification];
+  }
 
   [[NotifeeCoreUNUserNotificationCenter instance] observe];
 }
diff --git a/node_modules/@notifee/react-native/ios/NotifeeCore/NotifeeCore+UNUserNotificationCenter.m b/node_modules/@notifee/react-native/ios/NotifeeCore/NotifeeCore+UNUserNotificationCenter.m
index beaa360..ff53bef 100644
--- a/node_modules/@notifee/react-native/ios/NotifeeCore/NotifeeCore+UNUserNotificationCenter.m
+++ b/node_modules/@notifee/react-native/ios/NotifeeCore/NotifeeCore+UNUserNotificationCenter.m
@@ -21,7 +21,6 @@
 #import "NotifeeCoreUtil.h"
 
 @implementation NotifeeCoreUNUserNotificationCenter
-
 struct {
   unsigned int willPresentNotification : 1;
   unsigned int didReceiveNotificationResponse : 1;
@@ -73,12 +72,31 @@ struct {
 
 - (nullable NSDictionary *)getInitialNotification {
   if (_initialNotificationGathered && _initialNotificationBlock != nil) {
+
     // copying initial notification
     if (_initialNotification != nil &&
         [_initialNoticationID isEqualToString:_notificationOpenedAppID]) {
-      NSDictionary *initialNotificationCopy = [_initialNotification copy];
+
+      NSMutableDictionary *event = [NSMutableDictionary dictionary];
+      NSMutableDictionary *initialNotificationCopy = [_initialNotification mutableCopy];
+      initialNotificationCopy[@"initialNotification"] = @1;
+
       _initialNotification = nil;
-      _initialNotificationBlock(nil, initialNotificationCopy);
+
+      // Sets the return payload for getInitialNotification()
+     _initialNotificationBlock(nil, initialNotificationCopy);
+
+      // Prepare onForegroundEvent() payload
+      event[@"detail"] = [initialNotificationCopy copy];
+      if ([event[@"detail"][@"pressAction"][@"id"] isEqualToString:@"default"]) {
+          event[@"type"] = @1;  // PRESS
+      } else {
+          event[@"type"] = @2;  // ACTION_PRESS
+      }
+
+      // Call onForegroundEvent() on Javascript side with payload:
+      [[NotifeeCoreDelegateHolder instance] didReceiveNotifeeCoreEvent:event];
+
     } else {
       _initialNotificationBlock(nil, nil);
     }
@@ -104,7 +122,6 @@ struct {
   NSDictionary *notifeeNotification =
       notification.request.content.userInfo[kNotifeeUserInfoNotification];
 
-  // we only care about notifications created through notifee
   if (notifeeNotification != nil) {
     UNNotificationPresentationOptions presentationOptions = UNNotificationPresentationOptionNone;
     NSDictionary *foregroundPresentationOptions =
@@ -148,23 +165,16 @@ struct {
       presentationOptions |= UNNotificationPresentationOptionAlert;
     }
 
-    NSDictionary *notifeeTrigger = notification.request.content.userInfo[kNotifeeUserInfoTrigger];
-    if (notifeeTrigger != nil) {
-      // post DELIVERED event
-      [[NotifeeCoreDelegateHolder instance] didReceiveNotifeeCoreEvent:@{
+    // post DELIVERED event
+    [[NotifeeCoreDelegateHolder instance] didReceiveNotifeeCoreEvent:@{
         @"type" : @(NotifeeCoreEventTypeDelivered),
         @"detail" : @{
           @"notification" : notifeeNotification,
         }
-      }];
-    }
+    }];
 
     completionHandler(presentationOptions);
 
-  } else if (_originalDelegate != nil && originalUNCDelegateRespondsTo.willPresentNotification) {
-    [_originalDelegate userNotificationCenter:center
-                      willPresentNotification:notification
-                        withCompletionHandler:completionHandler];
   }
 }
 

```

LunatiqueCoder committed:
fix(ios): Remote issues: #1109-#1096 onForegroundEvent() and getInitialNotification()

LunatiqueCoder committed:
bug/#1109-#1096-remote-onForegroundEvent-getInitialNotification

LunatiqueCoder committed:
Merge remote-tracking branch 'origin/bug/#1109-#1096-remote-onForegroundEvent-getInitialNotification' into bug/#1109-#1096-remote-onForegroundEvent-getInitialNotification

mikehardy commented:
Hey @LunatiqueCoder 👋 ! Haven't hacked on things with you in a while, good to see you

I need to focus first on making sure Notifee is compatible with react-native 0.76 as I believe we have New Architecture incompatibilities even with the compat mode

But right after that I was planning on tackling all of these issues related to FCM delivery and events not firing and such - and it looks like you have beaten me to it with this PR

This is an area I really only want to change once (or...one more time) though so I need to think pretty deeply about how exactly this should work (should it consume the initial notification or not? how *exactly* should it interoperate with react-native-firebase? etc). So please be patient as it may be a little bit before I get to this, but it is pretty important to get right and I'm excited to see you've already thought through it with this patch here

LunatiqueCoder commented:
@mikehardy 🍺 Hehe, long time indeed. Happy to talk to you again 

I'm still tinkering, but I was very excited to see I might have fixed some Github issues while I was working on my stuff, especially when I saw you're a core contributor here as well. 

I'm also planning to open source an Expo config plugin for Notifee's iOS remote notifications. Basically it will follow these steps here: https://notifee.app/react-native/docs/ios/remote-notification-support

----


Anyway, back to our issues:

<br>

> For one specific question:
> 
> > This will also enable getInitialNotification() for iOS if the app was opened from a terminated state by pressing the notification. However, [the docs say it should be deprecated](https://notifee.app/react-native/docs/events#app-open-events). How are we supposed to handle this scenario without getInitialNotification()? Should we remove that part from the docs?
> 
> The app is started and brought to the foreground and I believe a Notifee event handler is called, to deliver the event?

<br>

While trying to understand the code, I noticed that `getInitialNotification()` should work by implementation, but [it was ALWAYS checking for **local** notifications, and never for **remote** ones](https://github.com/invertase/notifee/pull/1118/files#diff-e6cee9e1068ee77f7db69b7ee7ccfecefab70ddee393961f93eeeb034d4919afL61-L67).  

This is all I did to make it work: https://github.com/invertase/notifee/blob/5d46de0867f3e8fe5e0712547024186e3ebd695c/ios/NotifeeCore/NotifeeCore%2BNSNotificationCenter.m#L61-L72

----

As a side note, since **`apns-push-type: background`** it's almost useless, we had to rely on remote notifications for iOS without breaking Android. 

We ended up using this following snippet with Firebase Node.js SDK 
which I think it would be very useful to add in the docs): 

```ts
import type {NextApiRequest, NextApiResponse} from 'next';
import type {Notification} from '@notifee/react-native/src/types/Notification';
import {AndroidImportance} from '@notifee/react-native/src/types/NotificationAndroid';
import {MulticastMessage} from 'firebase-admin/lib/messaging/messaging-api';


const notifeeOptions: Notification = {
  id: Date.now().toString(),
  title: 'Title',
  subtitle: 'Subtitle',
  body: 'Main body content of the notification',
  android: {
    channelId: 'default',
    importance: AndroidImportance.HIGH,
    pressAction: 
  },
  ios: {
    sound: 'default',
    criticalVolume: 1,
    critical: true,
    foregroundPresentationOptions: {
      badge: true,
      banner: true,
      list: true,
      sound: true,
    },
  },
};



const message: MulticastMessage = {
  // ✅ We can continue using local notifcation for Android
  // 👍 while triggering iOS remote notifications from `apns`
  data: {body: JSON.stringify(notifeeOptions)}, 
  tokens: [],
  android: {
    priority: 'high',
  },
  apns: {
    payload: {
      notifee_options: notifeeOptions,
      aps: {
        alert: {
          // 🚧 This is needed to trigger an alert/remote notification only for iOS
          // 👍 but Android will continue using data-only notifications
          title: 'Hello World',
        },
        mutableContent: true,
      },
    },
  },
};

export default async (req: NextApiRequest, res: NextApiResponse) => {
  try {
    await admin.messaging().sendEachForMulticast(message);

    res.status(200).end();
  } catch (e) {
    res.status(400).end();
  }
};
```

<br>

dprevost-LMI commented:
FYI, I just tested the fix, and on my side, the first index in the patch differed from mine.

Also, after some JS debugging, the call to `await notified.getInitialNotification();` was not returning at all, aka the promise was hanging there.
I'm not an iOS expert, so something else I'm not aware from our project could interfere!

The different index after reconstructing the patch manually:
![image](https://github.com/user-attachments/assets/17d392dc-4147-4857-862c-f725dfee6850)

LunatiqueCoder commented:
Hmm, just noticed those changes were made with `@notifee/react-native@9.0.2`. If I update to the latest version, the patch fails. I'll try again and see how it goes. Thanks for the input!

dprevost-LMI commented:
I'm also on `"@notifee/react-native@9.0.2":`

dprevost-LMI commented on ios/NotifeeCore/NotifeeCore+NSNotificationCenter.m:
```diff
@@ -58,13 +58,18 @@ + (void)load {
-  [[NotifeeCoreUNUserNotificationCenter instance]
-      onDidFinishLaunchingNotification:launchNotification.userInfo];
-  [[NotifeeCoreUNUserNotificationCenter instance] getInitialNotification];
+  NSDictionary *notifUserInfo =
+      notification.userInfo[UIApplicationLaunchOptionsLocalNotificationKey];
+    
+  if (!notifUserInfo) {
+      // Fallback to remote notification key if local notification key is not available
+      notifUserInfo = notification.userInfo[UIApplicationLaunchOptionsRemoteNotificationKey];
+  }
+
+  if (notifUserInfo) {
+      [[NotifeeCoreUNUserNotificationCenter instance] onDidFinishLaunchingNotification:notifUserInfo];
+      [[NotifeeCoreUNUserNotificationCenter instance] getInitialNotification];
+  }
```
I ported just those changes in my project, and `await notifee.getInitialNotification();` hangs!
<img width="1261" alt="image" src="https://github.com/user-attachments/assets/55f44246-124f-4c7d-84b5-3c82818a4871">

```
    console.log('notifee.getInitialNotification-before');
    const initialNotification = await notifee.getInitialNotification();
    console.log('notifee.getInitialNotification-after'); <== not reached
```

LunatiqueCoder replied:
@dprevost-LMI  Hmm are you also using remote notifications? just double checking. Some debugger info from breakpoints would be great (that's how I solved my issue), but xcode debugging can be very hard if you never did it before. Learned it the hard way.

dprevost-LMI replied:
> are you also using remote notifications?

We were for sure since it used to work, but I just took over a colleague's branch work for the whole Android 14 deadline; he might have "remove it"! I'll dig!

LunatiqueCoder replied:
At first, I also started with that change, and managed to make it partially work. But eventually I concluded that changing `application_onDidFinishLaunchingNotification()` would be the most elegant solution that fixed lots of issues for me. Please keep me posted!

LunatiqueCoder replied:
@dprevost-LMI By the way, be sure to set `notifee_options` like this: 

```
const message: MulticastMessage = {
  // ✅ We can continue using local notifcation for Android
  // 👍 while triggering iOS remote notifications from `apns`
  data: {notifee_options: JSON.stringify(notifeeOptions)}, <---- Maybe this oone too, but `apns` is more important
  tokens: [],
  android: {
    priority: 'high',
  },
  apns: {
    payload: {
      notifee_options: notifeeOptions, // <---- Notifee needs this here
      aps: {
        alert: {
          // 🚧 This is needed to trigger an alert/remote notification only for iOS
          // 👍 but Android will continue using data-only notifications
          title: 'Hello World',
        },
        mutableContent: true,
      },
    },
  },
};
```

You can see this is needed here:

https://github.com/invertase/notifee/blob/main/ios/NotifeeCore/NotifeeCoreUtil.m#L625

and many other places like: 

https://github.com/invertase/notifee/blob/main/ios/NotifeeCore/NotifeeCoreExtensionHelper.m#L43-L49

mikehardy replied:
I'll resolve this per request, but that commit you linked (that you are - I assume - locally reverting via patch-package was a bit touchy and there were issues for at least one other person. So that area of code + that change in particular is still on my "problematic behavior" list. I'm not proprietary about that list of course, anyone that wants to think deeply and propose a change is welcome and may beat me to it 😅

dprevost-LMI replied:
I might try one day; I lack so much knowledge about that right now.
And yes, it is a patch locally! Mainly to meet Android 14 deadline dates!
But for sure, we will review it!

LunatiqueCoder replied:
---
> I tried to debug with Xcode on a working version, and even when using `Wait for the executable to be launched, the keys in question are nil.

1. That's why if **UIApplicationLaunchOptions<ins>Local</ins>NotificationKey** doesn't do the job, we try to use **UIApplicationLaunchOptions<ins>Remote</ins>NotificationKey** instead as a fallback. But if neither is working for you, I would expect you to hit # 3 in this list.

<br> 

----
> I cannot explain, but this is definitely related to the "remote" setup, where we have a second target named notifications that handles the reception and decryption of the notifications.

2. We're also using remote notifications by adding NotificationServiceExtension as a second target [as decribed in the docs](https://notifee.app/react-native/docs/ios/remote-notification-support). 

<br> 

---
> after reverting changes of commit [c9593cd](https://github.com/invertase/notifee/commit/c9593cded6dd98595a62d1eb9c2d23deb8cb445b) and testing on a release build, my background case and my close case started working.

3. In the beginning, I was also tempted to change this logic here, it really seemed to be the right thing to do, however, after finding # 1 in this list, I noticed I have the correct data flowing in `notifeeNotification` so no need to revert those for me anymore

dprevost-LMI replied:
Last update, I was tired and did not see this, but for the close case, finally, on top of the patch, we had to keep the event in memory after the `onForegroundEvent` or `onBackgroundEvent` (both call the same function). Then on boot, after the redux event rehydrated, we process the tapped notification!

LunatiqueCoder committed:
bug/#1109-#1096-remote-onForegroundEvent-getInitialNotification - call onForegroundEvent() when app starts from terminated state - same as getInitialNotification()

LunatiqueCoder commented:
Pushed one more commit to also fix `onForegroundEvent()` not being called when app is started from a terminated state. Now it's called after `getNotification()`

Perhaps there's a more elegant solution, and we could do it better. However, I believe the `didReceiveNotificationResponse()` method of `NotifeeCore+UNUserNotificationCenter.m` can be updated without this old time delay workaround:

https://github.com/invertase/notifee/commit/c606a85d730259fb418d7d64d5534a2113742ca9#diff-ec80c62b93aed0afa4addac8a6cd5fb6534543b3cc39bbd1b706adb9ec82a1ccL122-R132

https://github.com/invertase/notifee/blob/c606a85d730259fb418d7d64d5534a2113742ca9/ios/NotifeeCore/NotifeeCore%2BUNUserNotificationCenter.m#L123-L129

<br>

But for now it works (on my computer 🤣) and we can improve the logic over time. Of course, there's also the New Arch and so on. :) 

Cheers 🍻

LunatiqueCoder committed:
bug/#1109-#1096-remote-onForegroundEvent- update explanatory comment

LunatiqueCoder commented:
Hi again. Based on the conversation so far I updated the first post:

- Adapted the patch to use the latest Notifee version
- Added `What's changed` section describing the code changes decisions.
- Improved `How to test` section

LunatiqueCoder reviewed (commented):
Leaving a note for myself. From my testing, `onForeground` gets called twice for `DELIVERED` events on iOS

Mateo-Abribat commented:
Amazing work hope to see news about this PR soon 👀 thanks !

TestWts2017 commented:
when we can expect the latest version with the updated code?

mikehardy commented:
@TestWts2017 you can't. It is open source and I am always very clear with people - when you interact with open source you must abandon your expectations of release dates. If it is important to you, use patch-package and integrate the changes in your project directly, no need to wait

LunatiqueCoder commented:
☝️ the referenced issues above might be duplicated

mikehardy commented:
They probably are, most of the complaints in this repo are all actually the same complaint / disagreement with how it functions, or some ridiculously niche thing which is hard to solve 😆 

Been traveling all week but will be back to work next week

Mateo-Abribat commented:
Hey there! I partly agree with this because the Firebase documentation can be pretty unclear, especially when it comes to integration. It often feels like you have to dig through issues or PRs to get answers. Also, I wonder why Invertase promotes Notifee in react-native-firebase if it’s considered niche? Notifee is such a powerful library and a great addition to any React Native project. Ideally, it should work seamlessly on both iOS and Android with Firebase—if not, maybe it shouldn’t be featured as an option!

mikehardy commented:
Just to answer one part: Notifee is not at all considered niche! However, a full screen notification is a niche case, as is a permanent notification attached to a foreground service of a niche type, as is scheduling exact alarms. Each of those is a niche feature just based on query traffic here related to them. Non-niche use is "I want to post a notification, perhaps with styling and an image, in the foreground and background case"

It just so happens that this PR is right in that area of work.

However, my mention of the other stuff as niche is not to say Notifee shouldn't be supporting it. And in either case, react-native-firebase *had* to eject the local-notification device API surface area specifically because all those niche use cases (which are each important to someone!) were a huge drag on the core purpose of react-native-firebase which is to wrap firebase APIs.

It is not possible (we tried!) to support the local device notification APIs people want *and* all the firebase APIs in the react-native-firebase repos, we had to split it out, thus Notifee exists.

And unfortunately, with the pace of device-local notification API change (typically in the form of increasing restrictions to curb abusive behavior from app developers...) and the pace of change of device-local power management which affects message delivery (typically in the form of increasing restrictions to curb abusive behavior from app developers...) there is no simple integration. Troubleshooting unfortunately always starts with a tree of questions of the form "okay, iOS or Android? okay is it a notification, data, or mixed FCM? okay, what version of the operating system?" and gets worse from there.

There is no seemless unfortunately. The work in this PR should make it a great deal better though because yes, it is rough going right now.

Hope that's sort of clear with regard to maintainer goals and repo motivation at least?

Mateo-Abribat commented:
I completely agree that separating Notifee from react-native-firebase was the right approach! I also understand the difficulty of meeting OS requirements for both Android and iOS—each has its own set of complexities. Over the past few months, I’ve read through many Notifee issues and comments to better understand how notifications work across platforms, and I want to acknowledge and appreciate all the hard work put into this package.

I’m excited to see the improvements in the PR, as they’ll make these features more accessible. That said, I know that adding new functionality without impacting other areas is a significant challenge, especially given the wide range of use cases Notifee supports. Enhanced documentation, particularly around Firebase integration, could also help more developers avoid common pitfalls and make setup easier. I’ve noticed that this repo used to get frequent responses from you and other maintainers, so better documentation could help lighten that load too.

Thanks for the insight—it really clarifies the goals and challenges behind Notifee!

mikehardy commented:
I know there is a lot of interest in this one and I am as well. Just wanted to let everyone know it won't sit forever but I'm still backlogged on urgent items - that is there is one repo left that still does not seem to be behaving well with new architecture at all in the form of @invertase/react-native-apple-authentication and I must get that working first. cocoapods/xcodeproj blowing up all react-native-firebase builds last week was unexpected and took time as well.

There is one other new architecture issue here that came up where foreground services appear to freeze and I'll need to handle that as a regression before coming back to this, but once all the new arch stuff is done, this is first. It will happen but I'll appreciate the patience in the meantime

mikehardy commented:
Getting closer on this - apple auth qualified on new arch, and got reproduction scaffolding setup in react-native-firebase in order to test this stuff over the weekend https://github.com/invertase/react-native-firebase/pull/8143 - there is methodical progress towards having real test coverage for all the foreground / background stuff. I continue to appreciate the patience

dhruvvbhavsar commented:
Thanks man! This fix really helped!!!!
Keep doing the good man's work!

focux commented:
I'm wondering if this PR also fix the problem where you can't update the badge in the `onBackgroundEvent` handler... I'm asking because I see that this PR also fix some issues the background event.

TestWts2017 commented:
Great library!!!
But with the above patch package (tried with latest version also and mentioned version also) or with an older version, I am not able to get any payload for remote notification in iOS, but with Android, it is working. Any help or feedback is appreciated.

LunatiqueCoder commented:
@TestWts2017 [commented](https://github.com/invertase/notifee/pull/1118#issuecomment-2519503940):
> I am not able to get any payload response for remote notification in iOS, but with Android, it is working.
---- 


<details>
   <summary>
   👈 [click]   Did you try this sauce along with the patch?
   </summary>

<br>

 📖 **Source:** Answered above in https://github.com/invertase/notifee/pull/1118#issuecomment-2400737750


<br>

<img width="500" alt="image" src="https://github.com/user-attachments/assets/1183385f-7c99-4ffc-a992-02a8f3966741">

</details>

LunatiqueCoder commented:
@TestWts2017 on iOS, you will not receive anything in **onBackgroundEvent**. You will get the payload when you open the app in **onForegroundEvent** or **getInitialNotification**. Let me know if this works for you.

TestWts2017 commented:
> @TestWts2017 on iOS, you will not receive anything in `onBackgroundEvent`. You will get the payload when you open the app in `onForegroundEvent` or `getInitialNotification`. Let me know if this works for you.

Yes means `onForegroundEvent` is not able to log payload so is there any additional configuration we need to add

lucaritossa commented:
> Personally, I'm using shared group preferences to store background notifications[ within the notifications extension](https://notifee.app/react-native/docs/ios/remote-notification-support#2-in-a-notification-service-extension-in-your-app-when-a-device-receives-a-remote-message), and retrieve them on JS side with [`react-native-shared-group-preferences`](https://github.com/KjellConnelly/react-native-shared-group-preferences). That's how I get my background notifications on iOS

@LunatiqueCoder this sounds interesting but it is not so clear to me. Could you explain with more details? 
Thanks in adavance for any help!

ng-ha commented:
@LunatiqueCoder This patch make my app to crash when pressing on a local notification to open app in killed state. I'm using xcode 16.2,  "react-native": "0.74.3", "@notifee/react-native": "9.1.1"

mikehardy commented:
@ng-ha "....and the stack trace from the crash is...."

ng-ha commented:
@mikehardy Sorry for the delayed reply. I had to reproduce the crash. The previous crash was more informative, with the specific location of the crash. It came from NotifeeCore+UNUserNotificationCenter.m, but I didn't save that stack trace. Here is the stack trace I got this time with "@notifee/react-native": "^9.1.8". Thank you in advance!
[detail.txt](https://github.com/user-attachments/files/18240320/detail.txt)

ng-ha commented:
Here is the stack trace with "@notifee/react-native": "9.1.1" 
[detail-9.1.1.txt](https://github.com/user-attachments/files/18240478/detail-9.1.1.txt)

v-andreichuk commented:
Hi, really appreciate this library. I am writing to ask concerning listeners and maybe this PR will resolve the issue, but I'm not sure...

I have 2 events caught:
`[onMessage] (messaging API)::: {"contentAvailable": true, "data": {"chatId": "123"}, "from": "64154584446", "messageId": "1736823581544108", "notification": {"body": "SUPER TEST", "title": "TEST"}}`

`[notifee/onForegroundEvent] {"notification": {"body": "SUPER TEST", "data": {}, "id": "KK1V647KxyrxXKf19QQ8", "ios": {"foregroundPresentationOptions": [Object]}, "remote": false, "title": "TEST"}}`

As you can see - the data obj is empty in the notifee's event, and I wondered why. I tried to parse the payload using multiple approaches but it did not help.

Here is how I send the message from Node.js (`contentAvailable` is true, because without it - messaging().setBackgroundMessageHandler won't see this message):
`messaging
	.sendEachForMulticast({
		tokens: [ //tokens ],
		notification: {
			title: 'TEST',
			body: 'SUPER TEST',
		},
		apns: {
			payload: {
				aps: {
					contentAvailable: true,
				},
			},
		},
		data: { chatId: '123' },
	})`
	
Hope I can find answers here and thanks in advance!

stephanie-bassey commented:
tried the patch and the suggestion in [this comment](https://github.com/invertase/notifee/pull/1118#discussion_r1795086467), but I'm still facing an issue where launching the app from terminated via a quick action loses the notification payload entirely once the app is open. any suggestions?

andycrum commented:
Hi -- any possibility of an update here? The effect of this for us is that any push notifications delivered when the app is backgrounded have zero context when the app is opened from them -- and thus zero ability to link into specific screens within the app. Is there a workaround or is this any closer to getting resolved?

billnbell3 commented:
Does this need to be committed? On new ARCH we are having issues with IOS with FG mode.

al14z commented:
> @TestWts2017 on iOS, you will not receive anything in `onBackgroundEvent`. You will get the payload when you open the app in `onForegroundEvent` or `getInitialNotification`. Let me know if this works for you.
> 
> **EDIT:**
> 
> > Yes means `onForegroundEvent` is not able to log payload so is there any additional configuration we need to add
> 
> `onForegroundEvent` will get called if you open the app by interacting with the notification.
> 
> This is an iOS issue. It is not an issue with `react-native-firebase` or `notifee`. Apple will not let you run background logic because developers abuse it (like mining cryptocurrencies, eg bitcoin 😂)
> 
> Personally, I'm using shared group preferences to store background notifications[ within the notifications extension](https://notifee.app/react-native/docs/ios/remote-notification-support#2-in-a-notification-service-extension-in-your-app-when-a-device-receives-a-remote-message), and retrieve them on JS side with [`react-native-shared-group-preferences`](https://github.com/KjellConnelly/react-native-shared-group-preferences). That's how I get my background notifications on iOS

Has anyone else made @LunatiqueCoder's approach work? I have tried both using `getInitialNotification()` and [react-native-shared-group-preferences](https://github.com/KjellConnelly/react-native-shared-group-preferences) but can't seem to make it work - mind I'm in a React Native version prior to the architecture change.

rahulshettykcf commented:
> Hi -- any possibility of an update here? The effect of this for us is that any push notifications delivered when the app is backgrounded have zero context when the app is opened from them -- and thus zero ability to link into specific screens within the app. Is there a workaround or is this any closer to getting resolved?

This happens to me too on Android.
